### PR TITLE
Optimized persistence and rendering

### PR DESCRIPTION
**Pull request by matthewgertner**

### CODE REVIEW

File: assets/style.styl [hunk 1]
```diff
@@ -54,6 +54,10 @@ tour-tooltip
     height 0
     position absolute
 
+  // Hidden
+  &.inactive
+    display: none
+
```
Comment (matthewgertner) [inline assets/style.styl:60]: Unrelated fix since the hidden tooltip was causing UI issues since it was still layered above other elements in the UI.

File: src/Beacon.jsx [hunk 2]
```diff
@@ -13,6 +13,11 @@ const TARGET_CLONE_ID = 'tour-target-clone';
 // If margins are within tolerance then we center tooltip
 const TOOLTIP_TOLERANCE = 50;
 
+const HASH_CHECK_PENDING = 'PENDING';
+const HASH_CHECK_NOT_PERSISTENT = 'NOT_PERSISTENT';
+const HASH_CHECK_FOUND = 'FOUND';
+const HASH_CHECK_NOT_FOUND = 'NOT_FOUND';
```
Comment (matthewgertner) [inline src/Beacon.jsx:19]: Decided the way it was before was confusing and we should explicitly define all the possible states wrt the hash used for persistence.

File: src/Beacon.jsx [hunk 3]
```diff
@@ -44,17 +49,17 @@ export class Beacon extends Component {
...
   componentWillMount() {
     const inactive = (this.props.active === false) || (this.context.beacon && (this.context.beacon.active === false));
-    const persistent = this.props.persistent || (this.context.beacon && this.context.beacon.persistent);
-    const indexedDB = (this.context.beacon && this.context.beacon.indexedDB) || window.indexedDB;
-    this.setState({ inactive, persistent, indexedDB });
+    const persistent = this.context.beacon && this.context.beacon.persistent;
+    this.setState({ inactive, persistent });
 
-    this.checkHash();
+    this.loadHash(persistent);
```
Comment (matthewgertner) [inline src/Beacon.jsx:62]: So here we load the hash from the DB if appropriate.

File: src/Beacon.jsx [hunk 4]
```diff
@@ -77,8 +82,28 @@ export class Beacon extends Component {
     }
   }
 
+  shouldComponentUpdate(nextProps, nextState) {
```
Comment (matthewgertner) [inline src/Beacon.jsx:85]: This is an important optimization since all the states changes are causing a crazy about of re-rendering even though they often don't make any difference to the way the component is rendered (e.g. they affect how the tooltip is displayed when the beacon is clicked, but right now we are still showing the beacon).

File: src/Beacon.jsx [hunk 5]
```diff
@@ -77,8 +82,28 @@ export class Beacon extends Component {
...
+      return true;
+    } else {
+      // In all other cases we can optimize by skipping the update
+      return false;
+    }
+  }
+
   componentWillUpdate() {
-    this.checkHash();
+    const hashCheck = this.checkHashStatus();
```
Comment (matthewgertner) [inline src/Beacon.jsx:103]: So here we do a synchronous check to see if the hash status has changed in the cache. There isn't really any need to do an asynchronous database load here (thanks to the cache) and we need to be able to retrieve the hash status synchronously in `shouldComponentUpdate` as well.

File: src/Beacon.jsx [hunk 6]
```diff
@@ -77,8 +82,28 @@ export class Beacon extends Component {
...
   }
 
+  shouldComponentUpdate(nextProps, nextState) {
+    if (nextState.tooltip) {
+      // Always update if the tooltip is visible
+      return true;
+    } else if (Object.keys(this.props).some(key => this.props[key] !== nextProps[key])) {
+      // Update if any prop has changed
+      return true;
+    } else if (this.state.inactive !== nextState.inactive || this.state.hashCheck !== this.checkHashStatus()) {
```
Comment (matthewgertner) [inline src/Beacon.jsx:92]: See next comment.

File: src/Beacon.jsx [hunk 7]
```diff
@@ -99,14 +124,16 @@ export class Beacon extends Component {
...
-    if (boundsChanged || (this.state.targetElement !== targetElement)) {
-      this.setState({ targetElement, targetBounds }); // eslint-disable-line react/no-did-update-set-state
+    if (targetElement) {
+      const targetBounds = targetElement.getBoundingClientRect();
+      const boundsChanged =
+        this.state.targetBounds.top !== targetBounds.top ||
+        this.state.targetBounds.bottom !== targetBounds.bottom ||
+        this.state.targetBounds.left !== targetBounds.left ||
+        this.state.targetBounds.right !== targetBounds.right;
+      if (boundsChanged || (this.state.targetElement !== targetElement)) {
```
Comment (matthewgertner) [inline src/Beacon.jsx:134]: Trying to minimize the number of `setState` calls since it seems like even if I call it twice in the same lifecycle method, it still forces two rerenders (not 100% sure about that though).

File: src/Beacon.jsx [hunk 8]
```diff
@@ -177,8 +204,15 @@ export class Beacon extends Component {
     return targetClone;
   }
 
-  renderBeacon(persistent) {
-    if (this.state.inactive || (persistent && !this.state.hash)) {
+  getHash() {
```
Comment (matthewgertner) [inline src/Beacon.jsx:207]: Getter instead of using state for this (I'm sure you'll be a fan).

File: src/Beacon.jsx [hunk 9]
```diff
@@ -177,8 +204,15 @@ export class Beacon extends Component {
...
+    const { tooltipText, id } = this.props;
+    const text = typeof(tooltipText) === 'string' ? tooltipText : ReactDOMServer.renderToStaticMarkup(tooltipText);
```
Comment (matthewgertner) [inline src/Beacon.jsx:209]: If you can suggest a more robust way to do the `renderToStaticMarkup` thing then I'm all ears. I guess even if they move it to server-only code I can still include it somehow.

File: src/Beacon.jsx [hunk 10]
```diff
@@ -177,8 +204,15 @@ export class Beacon extends Component {
...
+    return id || sha1(text);
+  }
+
+  renderBeacon() {
+    const { inactive, hashCheck } = this.state;
+    if (inactive || (hashCheck === HASH_CHECK_PENDING) || (hashCheck === HASH_CHECK_FOUND)) {
```
Comment (matthewgertner) [inline src/Beacon.jsx:215]: If we found the hash that means the user has already clicked on the beacon.

File: src/Beacon.jsx [hunk 11]
```diff
@@ -220,7 +254,7 @@ export class Beacon extends Component {
       }
     }
 
-    const tooltipClass = tooltipActive ? 'tour-in' : !this.state.tooltipHidden && 'tour-out';
+    const tooltipClass = tooltipActive ? 'tour-in' : (!this.state.tooltipHidden && 'tour-out') || 'inactive';
```
Comment (matthewgertner) [inline src/Beacon.jsx:257]: Here we use that new `inactive` style to hide the tooltip after the user displays and then dismisses it.

File: src/BeaconConfig.jsx [hunk 12]
```diff
@@ -16,9 +16,38 @@ export default class BeaconConfig extends Component {
     indexedDB: indexedDB
   }
 
+  constructor(props) {
```
Comment (matthewgertner) [inline src/BeaconConfig.jsx:19]: So now `BeaconConfig` deals with all persistence stuff centrally.

File: src/BeaconConfig.jsx [hunk 13]
```diff
@@ -28,4 +57,33 @@ export default class BeaconConfig extends Component {
...
   }
+
+  storeHash(hash) {
+    const transaction = this.state.database.transaction(['beacons'], 'readwrite');
+    transaction.objectStore('beacons').add(true, hash);
+
+    this.setState({ dbCache: { ...this.state.dbCache, [hash]: hash }});
+  }
+
+  loadHash(hash, callback) {
```
Comment (matthewgertner) [inline src/BeaconConfig.jsx:68]: Callback is invoked with `true` if the hash is found (i.e. the user has already clicked on the beacon) and `false` otherwise.

File: src/BeaconConfig.jsx [hunk 14]
```diff
@@ -28,4 +57,33 @@ export default class BeaconConfig extends Component {
...
+    if (hash in this.state.dbCache && callback) {
+      callback(this.state.dbCache[hash] !== null);
+    } else if (!this.state.database) {
+      // Database is not open yet so queue request
+      this.state.requestQueue.push({ hash: hash, callback: callback });
+    } else {
+      const transaction = this.state.database.transaction(['beacons']);
+      const objectStore = transaction.objectStore('beacons');
+      objectStore.get(hash).onsuccess = (event) => {
+        this.state.dbCache[hash] = event.target.result ? hash : null;
```
Comment (matthewgertner) [inline src/BeaconConfig.jsx:78]: In the cache we store `null` if the has wasn't found.

File: src/BeaconConfig.jsx [hunk 15]
```diff
@@ -28,4 +57,33 @@ export default class BeaconConfig extends Component {
...
+        if (callback) {
+          callback(event.target.result);
+        }
+      };
+    }
+  }
+
+  checkHash(hash) {
+    return (hash in this.state.dbCache) && this.state.dbCache[hash];
```
Comment (matthewgertner) [inline src/BeaconConfig.jsx:87]: Returns `false` if we haven't cached the value, `null` if it wasn't found and the hash otherwise. These are then translated to the "enum" in `Beacon.jsx` although I kind of wonder if we shouldn't store those `HASH_CHECK_*` constants somewhere centrally and use them in both files.

File: src/Beacon.jsx [hunk 16]
```diff
@@ -77,8 +82,28 @@ export class Beacon extends Component {
     }
   }
 
+  shouldComponentUpdate(nextProps, nextState) {
+    if (nextState.tooltip) {
+      // Always update if the tooltip is visible
+      return true;
+    } else if (Object.keys(this.props).some(key => this.props[key] !== nextProps[key])) {
```
Comment (tomkis) [inline src/Beacon.jsx:89]: You know that this is something you never want to write on your own? https://github.com/gaearon/react-pure-render/blob/master/src/shallowEqual.js

I assume that the extra logic is probably mandatory.

File: src/Beacon.jsx [hunk 17]
```diff
@@ -77,8 +82,28 @@ export class Beacon extends Component {
     }
   }
 
+  shouldComponentUpdate(nextProps, nextState) {
+    if (nextState.tooltip) {
```
Comment (tomkis) [inline src/Beacon.jsx:86]: really? even if props or state changes?

File: src/Beacon.jsx (hunk 7 above)
Comment (tomkis) [inline src/Beacon.jsx:134]: that indeed doesn't make sense, unless you call it asynchronously.

File: src/Beacon.jsx [hunk 18]
```diff
@@ -239,55 +273,38 @@ export class Beacon extends Component {
...
-    const objectStore = transaction.objectStore('beacons');
-    objectStore.get(hash).onsuccess = (event) => {
-      if (!event.target.result) {
-        if (!this.state.hash) {
-          this.setState({ hash });
-        }
-      } else {
-        if (this.state.hash) {
-          this.setState({ hash: null });
+    if (persistent) {
```
Comment (tomkis) [inline src/Beacon.jsx:284]: Shoudln't you set `HASH_CHECK_PENDING` here ?

File: src/Beacon.jsx (hunk 9 above)
Comment (tomkis) [inline src/Beacon.jsx:209]: Yeah, no worries I guess it would be premature optimization anyway... I am nitpicking sometimes :P

File: src/Beacon.jsx (hunk 11 above)
Comment (tomkis) [inline src/Beacon.jsx:257]: Seriously, this is getting really difficult to reason about.

File: src/Beacon.jsx (hunk 18 above)
Comment (tomkis) [inline src/Beacon.jsx:284]: neverming... it's loading, not checking.

File: src/BeaconConfig.jsx (hunk 13 above)
Comment (tomkis) [inline src/BeaconConfig.jsx:68]: is there any particular reason that `loadHash` does not return a `Promise` ?

Comment (matthewgertner) [inline src/BeaconConfig.jsx:68]: It could be promise.
